packages/commerce-sdk-react/src/auth/index.ts

```diff
@@ -0,0 +1,339 @@
+            refresh_token: this.get('refresh_token_registered') || this.get('refresh_token_guest'),
+            token_type: this.get('token_type'),
+            usid: this.get('usid')
+        }
+    }
+
+    /**
+     * Used to validate JWT token expiration.
+     */
+    private isTokenExpired(token: string) {
+        if (!token) {
+            return true
+        }
+        const {exp, iat} = jwtDecode<JWTHeaders>(token.replace('Bearer ', ''))
+        const validTimeSeconds = exp - iat - 60
+        const tokenAgeSeconds = Date.now() / 1000 - iat
+        if (validTimeSeconds > tokenAgeSeconds) {
+            return false
+        }
+
+        return true
+    }
+
+    /**
+     * This function follows the public client auth flow to get
+     * access token for the user. Following the flow:
+     * 1. If we have valid access token - use it
+     * 2. If we have valid refresh token - refresh token flow
+     * 3. PKCE flow
+     *
+     * Only call this from the "ready" function, so "ready" can manage the pending state.
+     *
+     */
+    private async init() {
+        if (!this.isTokenExpired(this.get('access_token'))) {
+            return this.data
+        }
+
+        const refreshTokenRegistered = this.get('refresh_token_registered')
+        const refreshTokenGuest = this.get('refresh_token_guest')
+        const refreshToken = refreshTokenRegistered || refreshTokenGuest
+
+        if (refreshToken) {
+            try {
+                const res = await helpers.refreshAccessToken(this.client, {refreshToken})
+                this.handleTokenResponse(res, !!refreshTokenGuest)
+                return this.data
+            } catch {
+                // If anything bad happens during refresh token flow
+                // we continue with the PKCE guest user flow.
+            }
+        }
+
+        const res = await helpers.loginGuestUser(this.client, {redirectURI: this.redirectURI})
+        this.handleTokenResponse(res, true)
+
+        return this.data
+    }
+
+    /**
+     * This method stores the TokenResponse object retrived from SLAS, and
+     * store the data in storage.
+     */
+    private handleTokenResponse(res: ShopperLoginTypes.TokenResponse, isGuest: boolean) {
+        this.set('access_token', res.access_token)
+        this.set('customer_id', res.customer_id)
+        this.set('enc_user_id', res.enc_user_id)
+        this.set('expires_in', `${res.expires_in}`)
+        this.set('id_token', res.id_token)
+        this.set('idp_access_token', res.idp_access_token)
+        this.set('token_type', res.token_type)
+        this.set('usid', res.usid)
+
+        const refreshTokenKey = isGuest ? 'refresh_token_guest' : 'refresh_token_registered'
+        this.set(refreshTokenKey, res.refresh_token, {
+            expires: this.REFRESH_TOKEN_EXPIRATION_DAYS
+        })
+    }
+
+    /**
+     * The ready function returns a promise indicating whether we have
+     * a valid access token.
```

Comment:
```suggestion
     * The ready function returns a promise that resolves with a valid access token.
```
It does more than just indicate whether we have it, it gives it!